diff --git a/Eigen/src/Core/products/GeneralBlockPanelKernel.h b/Eigen/src/Core/products/GeneralBlockPanelKernel.h
index 3e69a26895..93798b00c6 100644
--- a/Eigen/src/Core/products/GeneralBlockPanelKernel.h
+++ b/Eigen/src/Core/products/GeneralBlockPanelKernel.h
@@ -1471,11 +1471,12 @@ EIGEN_DONT_INLINE void gebp_kernel<LhsScalar, RhsScalar, Index, DataMapper, mr,
         }
       }
 #endif
-      for (Index j2 = packet_cols8; j2 < packet_cols4; j2 += 4) {
+      tbb::parallel_for(packet_cols8 / 4, packet_cols4 / 4, [&](Index j2) {
+        j2 *= 4;
         for (Index i = i1; i < actual_panel_end; i += 2 * LhsProgress) {
           micro_panel(fix<2>, fix<4>, traits, i, j2);
         }
-      }
+      });
       for (Index j2 = packet_cols4; j2 < cols; j2++) {
         for (Index i = i1; i < actual_panel_end; i += 2 * LhsProgress) {
           micro_panel(fix<2>, fix<1>, traits, i, j2);
diff --git a/Eigen/src/Core/products/GeneralMatrixVector.h b/Eigen/src/Core/products/GeneralMatrixVector.h
index 40694f2e6f..75e93f5b02 100644
--- a/Eigen/src/Core/products/GeneralMatrixVector.h
+++ b/Eigen/src/Core/products/GeneralMatrixVector.h
@@ -221,7 +221,21 @@ general_matrix_vector_product<Index, LhsScalar, LhsMapper, ColMajor, ConjugateLh
   for (Index j2 = 0; j2 < cols; j2 += block_cols) {
     Index jend = numext::mini(j2 + block_cols, cols);
     Index i = 0;
-    for (; i < n8; i += ResPacketSize * 8) process_rows<8>(i, j2, jend, lhs, rhs, res, palpha, pcj);
+    for (; i < n8; i += ResPacketSize * 8)
+      for (auto x = 0; x < 8; ++x) {
+        const auto offset = i + ResPacketSize * x;
+        pstoreu(res + offset, pmadd(tbb::parallel_deterministic_reduce(
+                                        tbb::blocked_range<Index>(j2, jend), pzero(ResPacket{}),
+                                        [&](const tbb::blocked_range<Index>& r, ResPacket running_total) {
+                                          for (auto j = r.begin(); j < r.end(); ++j)
+                                            running_total += pcj.pmul(
+                                                lhs.template load<LhsPacket, LhsAlignment>(i + LhsPacketSize * x, j),
+                                                pset1<RhsPacket>(rhs(j, 0)));
+                                          return running_total;
+                                        },
+                                        std::plus<>()),
+                                    palpha, ploadu<ResPacket>(res + offset)));
+      }
     if (i < n4) {
       process_rows<4>(i, j2, jend, lhs, rhs, res, palpha, pcj);
       i += ResPacketSize * 4;
diff --git a/Eigen/src/SVD/BDCSVDImpl.h b/Eigen/src/SVD/BDCSVDImpl.h
index 4d0e70f77e..4cb67c775f 100644
--- a/Eigen/src/SVD/BDCSVDImpl.h
+++ b/Eigen/src/SVD/BDCSVDImpl.h
@@ -483,6 +483,16 @@ typename bdcsvd_impl<RealScalar_>::RealScalar bdcsvd_impl<RealScalar_>::secularE
                                                                                   const IndicesRef& perm,
                                                                                   const ArrayRef& diagShifted,
                                                                                   RealScalar shift) {
+  return tbb::parallel_deterministic_reduce(
+      tbb::blocked_range<Index>(Index(0), perm.size()), RealScalar(1),
+      [&](const tbb::blocked_range<Index>& r, RealScalar running_total) {
+        for (auto i = r.begin(); i < r.end(); ++i) {
+          const Index j = perm(i);
+          running_total += (col0(j) / (diagShifted(j) - mu)) * (col0(j) / (diag(j) + shift + mu));
+        }
+        return running_total;
+      },
+      std::plus<>());
   Index m = perm.size();
   RealScalar res = Literal(1);
   for (Index i = 0; i < m; ++i) {
@@ -815,6 +825,8 @@ template <typename RealScalar_>
 void bdcsvd_impl<RealScalar_>::computeSingVecs(const ArrayRef& zhat, const ArrayRef& diag, const IndicesRef& perm,
                                                const VectorType& singVals, const ArrayRef& shifts, const ArrayRef& mus,
                                                MatrixXr& U, MatrixXr& V) {
+  static tbb::affinity_partitioner ap;
+
   Index n = zhat.size();
   Index m = perm.size();
 
@@ -824,10 +836,13 @@ void bdcsvd_impl<RealScalar_>::computeSingVecs(const ArrayRef& zhat, const Array
       if (m_compV) V.col(k) = VectorType::Unit(n, k);
     } else {
       U.col(k).setZero();
-      for (Index l = 0; l < m; ++l) {
-        Index i = perm(l);
-        U(i, k) = zhat(i) / (((diag(i) - shifts(k)) - mus(k))) / ((diag(i) + singVals[k]));
-      }
+      tbb::parallel_for(
+          Index(0), m,
+          [&](const Index l) {
+            const Index i = perm(l);
+            U(i, k) = zhat(i) / (((diag(i) - shifts(k)) - mus(k))) / ((diag(i) + singVals[k]));
+          },
+          ap);
       U(n, k) = Literal(0);
       U.col(k).normalize();
 
